### DISCUSSION TITLE
PyG's Data.to(device) function behaves differently from PyTorch

### DISCUSSION THREAD

**Discussion by Boltzmachine** [pyg-team/pytorch_geometric#5433] [General]
In PyTorch, we now when `data` is a tensor, the `to(device)` function will return a copy, for example:
```python
>>> data = torch.tensor(1)
>>> data.device
device(type='cpu')
>>> data.to('cuda').device
device(type='cuda', index=0)
>>> data.device
device(type='cpu')
```
Howevew, if `data` is a PyG Data object, the `to(device)` function will inplace change the device of original data object, for example
```python
>>> data = Data(x=torch.tensor(1))
>>> data.x.device
device(type='cpu')
>>> data.to('cuda').x.device
device(type='cuda', index=0)
>>> data.x.device
device(type='cuda', index=0)
```

Is there any reason for such design. And what is the correct way if I want to make a cuda copy from PyG data in CPU.

**Comment by EdisonLeeeee** [comment]
Hi @Boltzmachine 

In PyG, `Data` is designed as a dictionary-like object, so `to(device)` operation is recursively applied to each key in `Data`. If you want to make a copy, you might use `copy.copy` like that:
```python
import copy
>>> data = Data(x=torch.tensor(1))
>>> data2 = copy.copy(data)

>>> data.to('cuda').x.device
device(type='cuda', index=0)
>>> data2.x.device
device(type='cpu')
```

**Reply by Boltzmachine** [reply]
Thanks so much! I also found this by checking the source code of InMemoryDataset.
